services/r2ctl/server/auth.go

```diff
@@ -0,0 +1,60 @@
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package server
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/gorilla/context"
+)
+
+const (
+	// UserIDField is the name of the variable that simple auth will put in the mux context
+	UserIDField = "userID"
+)
+
+// HTTPAuthService defines how to handle requests for various http authentication methods.
+type HTTPAuthService interface {
+	// User gets some identifying information about the user from the http request.
+	User(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc)
```

Comment:
does the signature have to be this for chaining?

Reply:
Yeah. https://github.com/urfave/negroni#handlers